generate-pdf.js

```diff
@@ -0,0 +1,97 @@
+const puppeteer = require("puppeteer");
+const yargs = require("yargs");
+
+DEFAULT_PRINT_OPTIONS = {
+  format: "Letter",
+  margin: {
+    top: "0.5in",
+    bottom: "0.5in",
+    right: "0.5in",
+    left: "0.5in"
+  }
+}
+
+GOTO_PDF_OPTIONS = {
+  waitUntil: "networkidle2"
+}
+
+VALIDATION_ERRORS = {
+  mustHaveWritepathOrRaw:
+    "At least one of the parameters { '--writePath <path>', '--raw' } must be supplied",
+  errorParsingJson:
+    "Error parsing option overrides as JSON: ",
+  urlIsMissing:
+    "The parameter '--url <url>' must be provided"
+}
+
+async function getUrlPdf(browser, url, writePath, optionOverrides) {
+  let pdfOptions = optionOverrides || DEFAULT_PRINT_OPTIONS;
+
+  pdfOptions.path = writePath;
+
+  const page = await browser.newPage();
+  await page.goto(url, GOTO_PDF_OPTIONS)
+  return await page.pdf(pdfOptions)
+}
+
+function validatePdfCliArgs(argv) {
+  let validationErrors = []; 
+
+  if (!argv.writePath && !argv.raw) {
+    validationErrors.push(VALIDATION_ERRORS.mustHaveWritepathOrRaw)
+  }
+
+  try {
+    argv.optionOverrides && JSON.parse(argv.optionOverrides);
+  } catch(exception) {
+    validationErrors.push(VALIDATION_ERRORS.errorParsingJson + exception.message);
+  }
+
+  if (!argv.url) {
+    validationErrors.push(VALIDATION_ERRORS.urlIsMissing);
+  }
+
+  if(validationErrors.length > 0) {
+    throw Error(validationErrors.join('\n'))
+  }
+
+  return true;
+}
+
+async function getUrlPdfCli() {
+  const argv = yargs
+    .option("url", {
+      alias: "u",
+      describe: "URL to generate PDF from"
+    })
+    .option("writePath", {
+      alias: "w",
+      describe: "path to filesystem location where generated PDF file should be created, if parameter is supplied"
+    })
+    .option("raw", {
+      alias: "r",
+      type: "boolean",
+      describe: "if present, contents of the PDF will be written to standard out"
+    })
+    .option("optionOverrides", {
```

Comment:
Typing json in the command line is pretty cumbersome. Can we look into other options, perhaps having yargs put all unknown arguments into the option overrides? Then you could just type them like regular CLI args. If that seems too unstructured, maybe only selecting unknown arguments with a certain prefix, or something like that.

Reply:
Thats a great idea! I'll look into it

Reply:
Good point, reading from a file sounds good.